scripts/tooling/rescan.c

```diff
@@ -0,0 +1,276 @@
+				continue;
+			}
+
+			if (((vid == match_vid) || (match_vid == 0xffff)) &&
+			    ((pid == match_pid) || (match_pid == 0xffff))) {
+
+				D(1, "Found %s/%s with vendor id %04lx and product id %04lx",
+				  PCI_DEVICES_PATH, dent->d_name, vid, pid);
+
+				snprintf(path, sizeof(path), "%s/%s", PCI_DEVICES_PATH,
+					 dent->d_name);
+
+				ret = cb(path, sizeof(path), data);
+				if (ret < 0) {
+					goto out;
+				}
+				counter += ret;
+			}
+		}
+	}
+
+	ret = counter;
+
+out:
+	closedir(dir);
+
+	return ret;
+}
+
+static int pcie_remove_cb(char *path, size_t path_size, void *data)
+{
+	int fd;
+	int ret;
+
+	(void)data;
+
+	strncat(path, "/remove", path_size);
+
+	fd = open(path, O_WRONLY);
+	if (fd < 0) {
+		E("Failed to open %s: %s", path, strerror(errno));
+		return -errno;
+	}
+
+	if (write(fd, "1", 1) < 0) {
+		E("Failed to write to %s: %s", path, strerror(errno));
+		ret = -errno;
+		goto out;
+	}
+
+	path[strlen(path) - 7] = '\0';
+	D(1, "Removed PCIe device %s", path);
+
+	ret = 1; /* success: count the number of devices removed */
+
+out:
+	close(fd);
+
+	return ret;
+}
+
+static int pcie_remove(void)
+{
+	return pcie_walk_sysfs(TENSTORRENT_PCI_VENDOR_ID, 0xffff, pcie_remove_cb, NULL);
+}
+
+static int pcie_count_cb(char *path, size_t path_size, void *data)
+{
+	(void)path;
+	(void)path_size;
+	(void)data;
+
+	return 1;
+}
+
+static int pcie_rescan(void)
+{
+	int fd;
+	int ret;
+
+	fd = open("/sys/bus/pci/rescan", O_WRONLY);
+	if (fd < 0) {
+		E("Failed to open %s: %s", "/sys/bus/pci/rescan", strerror(errno));
+		return -errno;
+	}
+
+	if (write(fd, "1", 1) < 0) {
+		E("Failed to write to %s: %s", "/sys/bus/pci/rescan", strerror(errno));
+		close(fd);
+		return -errno;
+	}
+
+	close(fd);
+
+	ret = pcie_walk_sysfs(TENSTORRENT_PCI_VENDOR_ID, 0xffff, pcie_count_cb, NULL);
+
+	if (ret >= 0) {
+		D(1, "Found %d Tenstorrent PCIe devices", ret);
+	}
+
+	return ret;
+}
+
+/**
+ * Rescan the PCIe bus for Tenstorrent devices.
+ *
+ * @param tt_dev_name Name of the Tenstorrent device to rescan.
+ * @return 0 on success, or a negative error code on failure.
+ */
+int rescan_pcie(const char *tt_dev_name)
+{
+	int ret, fd;
+
+	/* First, try to rescan the device using IOCTL */
+	struct tenstorrent_reset_device reset = {
+		.in.output_size_bytes = sizeof(reset.out),
+		.in.flags = TENSTORRENT_RESET_DEVICE_RESTORE_STATE,
+	};
+
+	fd = open(tt_dev_name, O_RDWR);
+	if (fd < 0) {
+		E("Failed to open device %s: %s", tt_dev_name, strerror(errno));
+		return -errno;
+	}
+
+	ret = ioctl(fd, TENSTORRENT_IOCTL_RESET_DEVICE, &reset);
+	if (ret < 0) {
+		E("Failed to reset device: %s", strerror(errno));
+		ret = -errno;
+		goto out;
+	}
+
+	ret = pcie_walk_sysfs(TENSTORRENT_PCI_VENDOR_ID, 0xffff, pcie_count_cb, NULL);
+
+	if (ret >= 0) {
+		D(1, "Found %d Tenstorrent PCIe devices", ret);
+		/* No need for full PCIe rescan */
+		goto out;
+	}
+
+	/* Now try PCIe rescan */
+	(void)pcie_remove();
+	ret = pcie_rescan();
+out:
+	close(fd);
+	return ret;
```

Comment:
This should go into a separate function, maybe `pcie_rescan_sysfs()`

Reply:
Are you asking for two functions to be exposed in `rescan.h`?

Reply:
two static functions. I guess if we keep the sysfs one around, then we can at least rescan if the device isn't already visible.